File: Leetcode/C++/234. Palindrome Linked List/solution.cpp
// https://leetcode.com/problems/palindrome-linked-list/

#include <vector>

using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        vector<int> listNode;
        // Move linkedList to vector
        while (head)
        {
            listNode.push_back(head->val);
            head = head->next;
        }

        int left = 0, right = listNode.size() - 1;
        // Check palindrome
        while (left < right)
        {
            if (listNode[left] != listNode[right])
            {
                return false;
            }
            left += 1;
            right -= 1;
        }
        return true;
    }
};

class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        ListNode *slow = head, *fast = head, *reverse = NULL, *temp = head;
        while (fast && fast->next) // Move slow and reverse to center
        {
            fast = fast->next->next;
            temp = slow;
            slow = slow->next;
            temp->next = reverse;
            reverse = temp;
        }
        if (fast) // If fast is not NULL ListNode length is odd so move slow to next node
        {
            slow = slow->next;
        }
        while (slow) // Check palindrome
        {
            if (slow->val != reverse->val)
            {
                return false;
            }
            slow = slow->next;
            reverse = reverse->next;
        }
        return true;
    }
};

File: Leetcode/Java/234. Palindrome Linked List/solution.java
// https://leetcode.com/problems/palindrome-linked-list/description/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // When listnode has odd elements
        if (fast != null){
            slow = slow.next;
        }

        ListNode prev = null;
        while (slow != null) {
            ListNode nextNode = slow.next;
            slow.next = prev;
            prev = slow;
            slow = nextNode;
        }
        
        while (prev != null){
            if (head.val != prev.val){
                return false;
            }

            head = head.next;
            prev = prev.next;
        }

        return true;
    }
}